core/src/main/java/org/apache/iceberg/actions/RewriteDataFilesCommitManager.java

```diff
@@ -0,0 +1,223 @@
+          rewriteGroups, e);
+      throw e;
+    } catch (Exception e) {
+      LOG.error("Cannot commit groups {}, attempting to clean up written files", rewriteGroups, e);
+      rewriteGroups.forEach(this::abortFileGroup);
+      throw e;
+    }
+  }
+
+  /**
+   * An async service which allows for committing multiple file groups as their rewrites complete. The service also
+   * allows for partial-progress since commits can fail. Once the service has been closed no new file groups should not
+   * be offered.
+   * @param rewritesPerCommit number of file groups to include in a commit
+   * @return the service for handling commits
+   */
+  public CommitService service(int rewritesPerCommit) {
+    return new CommitService(rewritesPerCommit);
+  }
+
+  public class CommitService implements Closeable {
+    private final ExecutorService committerService;
+    private final ConcurrentLinkedQueue<RewriteFileGroup> completedRewrites;
+    private final List<RewriteFileGroup> committedRewrites;
+    private final int rewritesPerCommit;
+    private final AtomicBoolean running = new AtomicBoolean(false);
+
+    CommitService(int rewritesPerCommit) {
+      LOG.info("Creating commit service for table {} with {} groups per commit", table, rewritesPerCommit);
+      this.rewritesPerCommit = rewritesPerCommit;
+
+      committerService = Executors.newSingleThreadExecutor(new ThreadFactoryBuilder()
+          .setNameFormat("Committer-Service")
+          .build());
+
+      completedRewrites = Queues.newConcurrentLinkedQueue();
+      committedRewrites = Lists.newArrayList();
+    }
+
+    /**
+     * Starts a single threaded executor service for handling file group commits.
+     */
+    public void start() {
+      Preconditions.checkState(running.compareAndSet(false, true), "Rewrite Commit service already started");
+      LOG.info("Starting commit service for {}", table);
+      // Partial progress commit service
+      committerService.execute(() -> {
+        while (running.get() || completedRewrites.size() > 0) {
+          try {
+            if (completedRewrites.size() == 0) {
+              // Give other threads a chance to make progress
+              Thread.sleep(100);
+            }
+          } catch (InterruptedException e) {
+            Thread.currentThread().interrupt();
+            throw new RuntimeException("Interrupted while processing commits", e);
+          }
+
+          // Either we have a full commit group, or we have completed writing and need to commit what is left over
+          if (completedRewrites.size() >= rewritesPerCommit || (!running.get() && completedRewrites.size() > 0)) {
+            Set<RewriteFileGroup> batch = Sets.newHashSetWithExpectedSize(rewritesPerCommit);
+            for (int i = 0; i < rewritesPerCommit && !completedRewrites.isEmpty(); i++) {
+              batch.add(completedRewrites.poll());
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.



**Problem**  
 You are using a `ConcurrentLinkedQueue`, but your usage of `isEmpty()` and `poll()` may not be thread-safe at lines: **163 and 164**. If the `ConcurrentLinkedQueue` contains only **one element** and two threads perform this **same check at the same time** , the second thread&#39;s call to `poll()` will get `null`. This `null` may be used at line: **164**.

**Fix**  
 Consider calling `poll()`, checking instead of your current check if the returned object is `null`, and only then using that object.